matmul performance benchmarks

aocsa opened the pull request:
Minor PR following up the previous PR about sparse benchmarking utils https://github.com/pytorch/pytorch/pull/48397

Fixes #44634:  Performance benchmarks for matrix-matrix and matrix-vector ops (dense-sparse, sparse-sparse, and compare to dense-dense)

I ran  all benchmarks on an 2xRTX8000 machine with AMD 2970WX 24-cores for `DLMC/magnitude_pruning` dataset with different sparsity levels.  

---
<details><summary> forward tests (expand for details).
</summary>

- `sparse@sparse`
```
[------------------------------- cpu:matmul-forward -------------------------------]
                           |   0.5   |   0.7   |   0.8   |   0.9   |  0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense    |  108.1  |  100.5  |  101.3  |  108.4  |  98.4  |  187.4
      torch:sparse@sparse  |  659.1  |  368.8  |  156.5  |   53.3  |  26.8  |   14.9
      scipy:sparse@sparse  |  565.1  |  233.9  |  130.2  |   23.1  |  21.6  |   15.2

Times are in milliseconds (ms).

[----------------------------------- cuda:matmul-forward -----------------------------------]
                           |    0.5    |    0.7    |   0.8    |   0.9    |   0.95   |   0.98 
1 threads: ----------------------------------------------------------------------------------
      torch:dense@dense    |   2243.5  |   4392.5  |  4419.8  |  2272.3  |  4433.9  |  8920.1
      torch:sparse@sparse  |  21369.2  |  11877.6  |  7339.2  |  1787.2  |  1335.1  |   845.7

Times are in microseconds (us).

```
- `sparse@dense`
```
[------------------------------- cpu:matmul-forward -------------------------------]
                          |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense   |  105.8  |  103.8  |  103.0  |  104.4  |  104.4  |  197.0
      torch:sparse@dense  |  119.9  |  102.4  |   84.0  |   19.7  |   16.8  |   11.6
      scipy:sparse@dense  |  906.5  |  799.6  |  697.8  |  182.2  |  165.5  |  135.4

Times are in milliseconds (ms).

[------------------------- cuda:matmul-forward --------------------------]
                          |  0.5  |  0.7  |  0.8  |  0.9  |  0.95  |  0.98
1 threads: ---------------------------------------------------------------
      torch:dense@dense   |  2.2  |  4.4  |  4.4  |  2.3  |  4.5   |  2.3 
      torch:sparse@dense  |  5.7  |  6.6  |  4.5  |  1.4  |  1.4   |  1.3 

Times are in milliseconds (ms).

```
- `sparse@vector`
```
[----------------------------------- cpu:matmul-forward ----------------------------------]
                           |    0.5    |   0.7    |   0.8    |   0.9    |   0.95   |   0.98
1 threads: --------------------------------------------------------------------------------
      torch:dense@vector   |    510.6  |   505.8  |   759.6  |   782.1  |   682.4  |  764.6
      torch:sparse@vector  |  10122.8  |  6241.1  |  7935.6  |  2076.3  |  1049.5  |  826.3
      scipy:sparse@vector  |   1756.7  |  1033.9  |   678.2  |   343.5  |   168.5  |   65.4

Times are in microseconds (us).

[-------------------------------- cuda:matmul-forward --------------------------------]
                           |   0.5    |   0.7    |   0.8   |   0.9   |   0.95  |   0.98
1 threads: ----------------------------------------------------------------------------
      torch:dense@vector   |    36.1  |    21.5  |   21.6  |   21.5  |   21.6  |   21.5
      torch:sparse@vector  |  1099.2  |  1289.4  |  775.7  |  327.1  |  285.4  |  274.0

Times are in microseconds (us).

```
</details>


---
<details><summary> backward tests (expand for details).
</summary>

- `sparse@sparse`
```
[--------------------------------- cpu:matmul-backward ---------------------------------]
                           |   0.5    |   0.7    |   0.8    |   0.9    |   0.95  |   0.98
1 threads: ------------------------------------------------------------------------------
      torch:dense@dense    |   246.1  |   315.0  |   306.9  |   168.6  |  290.6  |  146.9
      torch:sparse@sparse  |  6417.5  |  4393.7  |  3012.7  |  1029.4  |  908.0  |  650.7

Times are in microseconds (us).

[----------------------------- cuda:matmul-backward -----------------------------]
                           |   0.5   |   0.7   |   0.8   |  0.9   |  0.95  |  0.98
1 threads: -----------------------------------------------------------------------
      torch:dense@dense    |    6.7  |   13.3  |   13.3  |   6.9  |  13.5  |   6.9
      torch:sparse@sparse  |  143.7  |  143.4  |  119.6  |  29.5  |  29.1  |  10.9

Times are in microseconds (us).

```
- `sparse@dense`
```
 [------------------------------ cpu:matmul-backward -------------------------------]
                          |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense   |  185.9  |  304.8  |  305.8  |  169.9  |  308.7  |  168.4
      torch:sparse@dense  |  407.9  |  345.8  |  274.6  |  114.2  |  163.6  |  230.5

Times are in milliseconds (ms).

[--------------------------- cuda:matmul-backward --------------------------]
                          |  0.5   |  0.7   |  0.8   |  0.9  |  0.95  |  0.98
1 threads: ------------------------------------------------------------------
      torch:dense@dense   |   6.7  |  13.3  |  13.3  |  6.9  |  13.4  |   6.9
      torch:sparse@dense  |  16.7  |  19.0  |  15.1  |  6.3  |   8.2  |  12.7

Times are in milliseconds (ms).

```  
</details>

Kindly review this  PR. cc @mruberry, @ngimel

aocsa committed:
first commit

first commit

bench matmul

aocsa committed:
add docs


docs updated

robieta commented on benchmarks/sparse/dlmc/spmm.py:
```diff
@@ -0,0 +1,150 @@
+    measurements = []
+
+    for i, timer in enumerate(timers * repeats):
+        m = timer.blocked_autorange(min_run_time=0.05)
```
Is this enough to get stable measurements?

aocsa replied:
For the maximum problem size, `hidden_size=2048`,  is not enough. I changed it to 1 by default which should be sufficient to be stable and this now can be adjusted in the terminal. 
`    parser.add_argument('--timer_min_run_time', default=1, type=int)`

robieta commented on benchmarks/sparse/dlmc/spmm.py:
```diff
@@ -0,0 +1,150 @@
+
+    for i, timer in enumerate(timers * repeats):
+        m = timer.blocked_autorange(min_run_time=0.05)
+        serialized_results.append(pickle.dumps(m))
```
Why do we need to pickle the intermediate results?

aocsa replied:
ohhh is not needed, I removed it.  I guessed that by modifying the metadata structure   the `benchmark_utils.Compare` would not work, but actually it works well.

robieta commented on benchmarks/sparse/dlmc/spmv.py:
```diff
@@ -0,0 +1,113 @@
+                        help='dataset name (random_pruning, variational_dropout, magnitude_pruning, \
+                            extended_magnitude_pruning)',
+                        default='random_pruning')
+    parser.add_argument('--operation',
```
This appears to be unused.

robieta reviewed (changes requested):
At a high level it seems like `spmm.py` and `spmv.py` could be merged by having a central definition of tasks
```
_TASKS = {
    "mm": (
        ("torch", "torch.mm(dx, dy)"),
        ("torch.sparse", "torch.sparse.mm(x, y)"),
        ("scipy", "scipy_coo_matmul(sx, sy)"),
    ),
    "mv": (
        ("torch", "torch.matmul(dx, vector)"),
        ("torch.sparse", "torch.matmul(x, vector)"),
        ("scipy", "scipy_coo_matmul(sx, vector)"),
    ),
    "backward": (
        ("torch", "torch_backward(dx, dy)"),
        ("torch.sparse", "sparse_torch_backward(x, y)"),
    )
}
```

and some switching code at the start:
```
tasks = _TASKS[operation]
if device != "cpu":
    # SciPy is CPU only.
    tasks = tuple(t for t in tasks if t[0] != "scipy")

loader = (
    load_spmv_dataset if operation == "mv" else
    functools.partial(load_spmm_dataset, spmm_type=spmm_type)
)
```

There's also a bit of logic for globals, but it doesn't seem insurmountable. WDYT?

mruberry commented on benchmarks/sparse/dlmc/README.md:
```diff
@@ -0,0 +1,20 @@
+# Sparse benchmarks
+
+These sets of benchmarks are for the sparse matrix functionality using a popular real dataset collection called 
```
This README is really cool.

robieta replied:
Agreed!

ngimel commented on benchmarks/sparse/dlmc/spmm.py:
```diff
@@ -0,0 +1,150 @@
+    a_dense.requires_grad = True
+    b_dense.requires_grad = True
+    r1 = a_dense.matmul(b_dense)
+    f1 = torch.sum(r1)
```
when you are doing `sum` in the function being benchmarked, it alters benchmarking results. You have to pass gradOutput to this function, and then do `r1.backward(gradOutput)`. Same for `sparse_torch_backward` below

ngimel commented on benchmarks/sparse/dlmc/spmm.py:
```diff
@@ -0,0 +1,150 @@
+                        default='random_pruning')
+    parser.add_argument('--operation',
+                        type=str,
+                        help='matmul or backward',
```
It's awkward naming, you should have `matmul` and `matmul_backward` e.g.

ngimel commented on benchmarks/sparse/dlmc/spmm.py:
```diff
@@ -0,0 +1,150 @@
+                ("matmul", device, "torch", "torch.mm(dx, dy)"),
+                ("matmul", device, "torch.sparse", "torch.sparse.mm(x, y)"),
+            ]
+    else:
```
check that it's supported value, otherwise any unexpected `operation` value would be interpreted as `backward`

ngimel commented on benchmarks/sparse/dlmc/spmv.py:
```diff
@@ -0,0 +1,113 @@
+    parser.add_argument('--operation',
+                        type=str,
+                        help='matmul',
+                        default='matmul')
```
you are not providing functions to benchmark backward, is it because backward is not implemented for spmv?

aocsa replied:
Yes, it's seems there is an issue for backward for `torch.matmul` when the input is `spase @ vector`.  Moreover there is no test for this case in the `test_autograd.py`

ngimel commented on benchmarks/sparse/dlmc/spmv.py:
```diff
@@ -0,0 +1,113 @@
+    print("== Formatted " + "=" * 80 + "\n" + "/" * 93 + "\n")
+    comparison.trim_significant_figures()
+    comparison.colorize()
+    comparison.print()
```
why are you printing both formatted and unformatted versions?

aocsa replied:
Not needed actually, so I removed the formatted version.

ngimel commented on benchmarks/sparse/dlmc/spmv.py:
```diff
@@ -0,0 +1,113 @@
+
+    for i, timer in enumerate(timers * repeats):
+        m = timer.blocked_autorange(min_run_time=0.05)
+        serialized_results.append(pickle.dumps(m))
```
As in spmm, no need to pickle intermediates

ngimel commented on benchmarks/sparse/dlmc/spmv.py:
```diff
@@ -0,0 +1,113 @@
+    table = [(m.task_spec.sub_label, m.task_spec.description,
+              m.metadata["device"], m.mean) for m in measurements]
+    df = pd.DataFrame(table, columns=['method', 'sparsity', 'device', 'time'])
+    df.to_pickle(df_output_path)
```
why to you want to serialize DataFrame and not comparison? comparison contains the same data. Also, probably disk serialization should be optional.

aocsa replied:
Not needed actually, I removed this

ngimel commented on benchmarks/sparse/dlmc/utils.py:
```diff
@@ -0,0 +1,141 @@
+def scipy_coo_matmul(mat1, mat2):
+    if isspmatrix(mat1) and isspmatrix(mat2):
+        result = mat1.dot(mat2).tocoo()
+        return torch.sparse_coo_tensor([result.row, result.col], result.data,
```
why are you converting result to torch coo tensor? Would that distort scipy benchmarking results?

aocsa replied:
Not needed actually, I simplified this to `mat1.dot(mat2)`

ngimel commented on benchmarks/sparse/dlmc/utils.py:
```diff
@@ -0,0 +1,141 @@
+        return torch.sparse_coo_tensor([result.row, result.col], result.data,
+                                       result.shape)
+    else:
+        return mat1.dot(mat2)
```
When is this path triggered? For mv? Would mat2 (that is a torch tensor in the mv call) be implicitly converted to ndarray?

aocsa replied:
All variables are converted to the right format before calling these functions.

mruberry commented on benchmarks/sparse/dlmc/utils.py:
```diff
@@ -0,0 +1,141 @@
+
+
+def read_matrix_params(path):
+    sys.stdin = open(path)
```
Should this open the path with a file descriptor and a context manager instead of assigning to sys.stdin?

mruberry commented on benchmarks/sparse/dlmc/utils.py:
```diff
@@ -0,0 +1,141 @@
+    return torch.tensor([rows, cols], dtype=torch.long)
+
+def load_sparse_matrix(path, device):
+    sys.stdin = open(path)
```
Same question here about reading from the path

ngimel commented:
Do you have results from running this benchmark?

aocsa commented:
> At a high level it seems like `spmm.py` and `spmv.py` could be merged by having a central definition of tasks
> 
> ```
> _TASKS = {
>     "mm": (
>         ("torch", "torch.mm(dx, dy)"),
>         ("torch.sparse", "torch.sparse.mm(x, y)"),
>         ("scipy", "scipy_coo_matmul(sx, sy)"),
>     ),
>     "mv": (
>         ("torch", "torch.matmul(dx, vector)"),
>         ("torch.sparse", "torch.matmul(x, vector)"),
>         ("scipy", "scipy_coo_matmul(sx, vector)"),
>     ),
>     "backward": (
>         ("torch", "torch_backward(dx, dy)"),
>         ("torch.sparse", "sparse_torch_backward(x, y)"),
>     )
> }
> ```
> 
> and some switching code at the start:
> 
> ```
> tasks = _TASKS[operation]
> if device != "cpu":
>     # SciPy is CPU only.
>     tasks = tuple(t for t in tasks if t[0] != "scipy")
> 
> loader = (
>     load_spmv_dataset if operation == "mv" else
>     functools.partial(load_spmm_dataset, spmm_type=spmm_type)
> )
> ```
> 
> There's also a bit of logic for globals, but it doesn't seem insurmountable. WDYT?

Indeed, it was surmountable. Not `spmm.py` and `spmv.py` are merged :)

aocsa committed:
feedback addressed 


update


update


update 2

aocsa commented:
> Do you have results from running this benchmark?

I addressed the feedback comments. Moreover,  I ran  all benchmarks on an 2xRTX8000 machine with AMD 2970WX 24-cores for `DLMC/magnitude_pruning` dataset with different sparsity levels.  

---
<details><summary> forward tests (expand for details).
</summary>

- `sparse@sparse`
```
[-------------------------------- cpu:matmul-forward -------------------------------]
                           |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: --------------------------------------------------------------------------
      torch:dense@dense    |  106.9  |  106.9  |   97.6  |  104.7  |  104.3  |  197.8
      torch:sparse@sparse  |  698.5  |  457.1  |  230.4  |   53.7  |   25.3  |   15.7
      scipy:sparse@sparse  |  553.0  |  221.9  |  114.0  |   22.5  |   18.3  |   11.4

Times are in milliseconds (ms).

[----------------------------------- cuda:matmul-forward -----------------------------------]
                           |    0.5    |    0.7    |   0.8    |   0.9    |   0.95   |   0.98 
1 threads: ----------------------------------------------------------------------------------
      torch:dense@dense    |   2243.5  |   4392.5  |  4419.8  |  2272.3  |  4433.9  |  8920.1
      torch:sparse@sparse  |  21369.2  |  11877.6  |  7339.2  |  1787.2  |  1335.1  |   845.7

Times are in microseconds (us).

```
- `sparse@dense`
```
[------------------------------- cpu:matmul-forward -------------------------------]
                          |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense   |  105.8  |  103.8  |  103.0  |  104.4  |  104.4  |  197.0
      torch:sparse@dense  |  119.9  |  102.4  |   84.0  |   19.7  |   16.8  |   11.6
      scipy:sparse@dense  |  906.5  |  799.6  |  697.8  |  182.2  |  165.5  |  135.4

Times are in milliseconds (ms).

[------------------------- cuda:matmul-forward --------------------------]
                          |  0.5  |  0.7  |  0.8  |  0.9  |  0.95  |  0.98
1 threads: ---------------------------------------------------------------
      torch:dense@dense   |  2.2  |  4.4  |  4.4  |  2.3  |  4.5   |  2.3 
      torch:sparse@dense  |  5.7  |  6.6  |  4.5  |  1.4  |  1.4   |  1.3 

Times are in milliseconds (ms).

```
- `sparse@vector`
```
[----------------------------------- cpu:matmul-forward ----------------------------------]
                           |    0.5    |   0.7    |   0.8    |   0.9    |   0.95   |   0.98
1 threads: --------------------------------------------------------------------------------
      torch:dense@vector   |    510.6  |   505.8  |   759.6  |   782.1  |   682.4  |  764.6
      torch:sparse@vector  |  10122.8  |  6241.1  |  7935.6  |  2076.3  |  1049.5  |  826.3
      scipy:sparse@vector  |   1756.7  |  1033.9  |   678.2  |   343.5  |   168.5  |   65.4

Times are in microseconds (us).

[-------------------------------- cuda:matmul-forward --------------------------------]
                           |   0.5    |   0.7    |   0.8   |   0.9   |   0.95  |   0.98
1 threads: ----------------------------------------------------------------------------
      torch:dense@vector   |    36.1  |    21.5  |   21.6  |   21.5  |   21.6  |   21.5
      torch:sparse@vector  |  1099.2  |  1289.4  |  775.7  |  327.1  |  285.4  |  274.0

Times are in microseconds (us).

```
</details>


---
<details><summary> backward tests (expand for details).
</summary>

- `sparse@sparse`
```
[--------------------------------- cpu:matmul-backward ---------------------------------]
                           |   0.5    |   0.7    |   0.8    |   0.9    |   0.95  |   0.98
1 threads: ------------------------------------------------------------------------------
      torch:dense@dense    |   246.1  |   315.0  |   306.9  |   168.6  |  290.6  |  146.9
      torch:sparse@sparse  |  6417.5  |  4393.7  |  3012.7  |  1029.4  |  908.0  |  650.7

Times are in microseconds (us).

[----------------------------- cuda:matmul-backward -----------------------------]
                           |   0.5   |   0.7   |   0.8   |  0.9   |  0.95  |  0.98
1 threads: -----------------------------------------------------------------------
      torch:dense@dense    |    6.7  |   13.3  |   13.3  |   6.9  |  13.5  |   6.9
      torch:sparse@sparse  |  143.7  |  143.4  |  119.6  |  29.5  |  29.1  |  10.9

Times are in microseconds (us).

```
- `sparse@dense`
```
 [------------------------------ cpu:matmul-backward -------------------------------]
                          |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: -------------------------------------------------------------------------
      torch:dense@dense   |  185.9  |  304.8  |  305.8  |  169.9  |  308.7  |  168.4
      torch:sparse@dense  |  407.9  |  345.8  |  274.6  |  114.2  |  163.6  |  230.5

Times are in milliseconds (ms).

[--------------------------- cuda:matmul-backward --------------------------]
                          |  0.5   |  0.7   |  0.8   |  0.9  |  0.95  |  0.98
1 threads: ------------------------------------------------------------------
      torch:dense@dense   |   6.7  |  13.3  |  13.3  |  6.9  |  13.4  |   6.9
      torch:sparse@dense  |  16.7  |  19.0  |  15.1  |  6.3  |   8.2  |  12.7

Times are in milliseconds (ms).

```  
</details>

cc @robieta, @ngimel, @mruberry

aocsa committed:
minor updates for performance tests


minor fix

mruberry reviewed (approved):
Cool! Thanks @aocsa. I wonder if we can close the gap with SciPy in the scenarios we're slower on CPU?

@robieta would you like to take another look before this is merged?

aocsa commented:
> Cool! Thanks @aocsa. I wonder if we can close the gap with SciPy in the scenarios we're slower on CPU?
> 
> @robieta would you like to take another look before this is merged?

Thanks @mruberry, the current implementation of sparse-sparse matmul has almost the same performance.  I forgot to considering the overhead of converting the `csr` output of the `x.dot(y)` scipy operation   to `coo` which is part of the process of `torch.sparse.mm` operation. I updated the results with this change.  

```
[-------------------------------- cpu:matmul-forward -------------------------------]
                           |   0.5   |   0.7   |   0.8   |   0.9   |   0.95  |   0.98
1 threads: --------------------------------------------------------------------------
      torch:dense@dense    |  111.2  |  104.9  |  111.9  |  108.6  |  100.0  |  192.1
      torch:sparse@sparse  |  707.6  |  347.1  |  196.7  |   75.7  |   25.7  |   15.1
      scipy:sparse@sparse  |  584.2  |  365.4  |  105.9  |   30.7  |   22.7  |   14.4
Times are in milliseconds (ms).
```
I moved main comment and benchmark results into the PR description.

aocsa committed:
update repeats and scipy_mm

update default timer_min_run_time


update


update


update repeats

mruberry commented:
Let's give @robieta a chance to review.

robieta commented on benchmarks/sparse/dlmc/matmul_bench.py:
```diff
@@ -0,0 +1,125 @@
+
+    dataset_path = args.path
+    dataset_name = args.dataset
+    dataset_path = f"{dataset_path}/{dataset_name}"
```
nit: `os.path.join(...)`

robieta reviewed (approved):
I had a single nitpick, but overall looks great. Thanks!

aocsa committed:
minor update

aocsa commented:
I rebased this PR and I think this  PR is ready to be merged. 
Thanks cc @mruberry

aocsa committed:
remove trailing spaces


quick-checks

mruberry commented on test/jit/test_remove_mutation.py:
```diff
@@ -7,7 +7,7 @@
 pytorch_test_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
 sys.path.append(pytorch_test_dir)
-from torch.testing._internal.jit_utils import JitTestCase, freeze_rng_state	
+from torch.testing._internal.jit_utils import JitTestCase, freeze_rng_state
```
I'm trying to land this but the internal tools are struggling to reconcile these whitespace changes. Would you revert these changes and/or rebase this PR, @aocsa?

aocsa replied:
Sure, it's done cc @mruberry
